### PR TITLE
#124 Should transform value sets lazily fix

### PR DESCRIPTION
**Pull request by chandresh-pancholi**

### CODE REVIEW

File: concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java [hunk 1]
```diff
@@ -170,5 +172,52 @@
...
+    public class LazyTransformSet<V1, V2> extends AbstractSet<V2>  {
+        private final Set<V1> original;
+        private final Function<V1, V2> function;
+
+        public LazyTransformSet(Set<V1> original, Function<V1, V2> function){
+            this.original = original;
+            this.function = function;
+        }
+
+        /**
```
Comment (jtnelson) [inline concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java]: Sorry, i should have clarified in my previous comment:

You don't need javadoc for overriden methods

File: concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java [hunk 2]
```diff
@@ -170,5 +172,52 @@
...
+                public V2 next() {
+                     V1 backingNext = backing.next();
+
+                     return function.apply(backingNext);
+                }
+            };
+        }
+
+        /**
+         * @return size of original set
```
Comment (jtnelson) [inline concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java]: see above about no javadoc for overriden methods

File: concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java [hunk 3]
```diff
@@ -170,5 +172,52 @@
...
+
+                @Override
+                public boolean hasNext() {
+                    return backing.hasNext();
+                }
+
+                @Override
+                public V2 next() {
+                     V1 backingNext = backing.next();
+
```
Comment (jtnelson) [inline concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java]: remove space

File: concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java [hunk 4]
```diff
@@ -170,5 +172,52 @@
...
     }
+    
+    /**
+     * LazyTransformSet class transport one set v1 into another set v2 lazily.
+     * 
+     * @author chandresh.pancholi
+     *
+     */
+    public class LazyTransformSet<V1, V2> extends AbstractSet<V2>  {
+        private final Set<V1> original;
```
Comment (jtnelson) [inline concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java]: need javadoc for instance variables

File: concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java [hunk 5]
```diff
@@ -170,5 +172,52 @@
         }
         return transformed;
     }
+    
+    /**
+     * LazyTransformSet class transport one set v1 into another set v2 lazily.
```
Comment (jtnelson) [inline concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java]: A better description:

```java
/**
 * A {@link Set} that transform values from an original set of type {@code V1} into values of type {@code V2}
 * on-the-fly using a transformation {@link Function}.
 */
```

In general, write javadoc as if you were writing it for a person reading the codebase for the first time. You want to provide helpful context that'll give them a foundation from which they can understand

**Comment by jtnelson** [comment]
Looks pretty good so far. Next do the following

1. Create a static factory method in the Transformers class `public static <V1, V2> Set<V2> transformSetLazily(Set<V1>, Function<V1, V2> function);`
2. Create unit tests for the `LazyTransformSet` class. At the very least, I would test the following scenario:
* Transform the same set using Transformers#transformSet and Transformers#transformSetLazily and test that the iterators for both the transformed sets returned the exact same values on each iteration.

File: concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java (hunk 5 above)
Comment (chandresh-pancholi) [inline concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java]: I am pretty bad with javadoc :( :( . I will definitely put more information.

File: concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java [hunk 6]
```diff
@@ -142,6 +144,27 @@
...
     }
+    /**
+     * Populate {@code transformed} with the items in {@code original} after
+     * applying Lazy transform
+     * @param original
+     * @param function
+     * @return the transformed Set
+     */
+    public static <V1, V2> Set<V2> transformSetLazily(Set<V1> original, Function<V1, V2> function) {
+        Set<V2> transformed = Sets.newLinkedHashSetWithExpectedSize(original
```
Comment (jtnelson) [inline concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java]: you don't need this

File: concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java [hunk 7]
```diff
@@ -142,6 +144,27 @@
...
+
+        LazyTransformSet<V1, V2> lazyTransformSet = (new Transformers()).
+                new LazyTransformSet<>(original, function);
+        
+        Iterator<V2> iterator = lazyTransformSet.iterator();
+        
+        while(iterator.hasNext()) {
+            transformed.add(iterator.next());
+        }
+        return transformed;
```
Comment (jtnelson) [inline concourse-driver-java/src/main/java/com/cinchapi/concourse/util/Transformers.java]: all you need to do in this method is
```java
return new LazyTransformSet<V1, V1>(original, function);
```

File: concourse-driver-java/src/test/java/com/cinchapi/concourse/util/TransformersTest.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        };
+        
+        Set<String> output = Transformers.transformSetLazily(original, function);
+        Set<String> transformedSetOutput = Transformers.transformSet(original, function);
+        
+        Iterator<String> it1 = output.iterator();
+        Iterator<String> it2 = transformedSetOutput.iterator();
+        
+        while(it1.hasNext() && it2.hasNext()) {
+            assertTrue(it1.next().equals(it2.next()));
```
Comment (jtnelson) [inline concourse-driver-java/src/test/java/com/cinchapi/concourse/util/TransformersTest.java]: use `assertEquals` instead of `assertTrue`

**Pull request merged**